subsys/bluetooth/audio/aics_client.c

```diff
@@ -0,0 +1,889 @@
+		struct bt_gatt_subscribe_params *sub_params = NULL;
+		struct bt_gatt_chrc *chrc;
+
+		chrc = (struct bt_gatt_chrc *)attr->user_data;
+		if (inst->cli.start_handle == 0) {
+			inst->cli.start_handle = chrc->value_handle;
+		}
+		inst->cli.end_handle = chrc->value_handle;
+
+		if (!bt_uuid_cmp(chrc->uuid, BT_UUID_AICS_STATE)) {
+			BT_DBG("Audio Input state");
+			inst->cli.state_handle = chrc->value_handle;
+			sub_params = &inst->cli.state_sub_params;
+		} else if (!bt_uuid_cmp(chrc->uuid, BT_UUID_AICS_GAIN_SETTINGS)) {
+			BT_DBG("Gain settings");
+			inst->cli.gain_handle = chrc->value_handle;
+		} else if (!bt_uuid_cmp(chrc->uuid, BT_UUID_AICS_INPUT_TYPE)) {
+			BT_DBG("Input type");
+			inst->cli.type_handle = chrc->value_handle;
+		} else if (!bt_uuid_cmp(chrc->uuid, BT_UUID_AICS_INPUT_STATUS)) {
+			BT_DBG("Input status");
+			inst->cli.status_handle = chrc->value_handle;
+			sub_params = &inst->cli.status_sub_params;
+		} else if (!bt_uuid_cmp(chrc->uuid, BT_UUID_AICS_CONTROL)) {
+			BT_DBG("Control point");
+			inst->cli.control_handle = chrc->value_handle;
+		} else if (!bt_uuid_cmp(chrc->uuid, BT_UUID_AICS_DESCRIPTION)) {
+			BT_DBG("Description");
+			inst->cli.desc_handle = chrc->value_handle;
+			if (chrc->properties & BT_GATT_CHRC_NOTIFY) {
+				sub_params = &inst->cli.desc_sub_params;
+			}
+
+			if (chrc->properties & BT_GATT_CHRC_WRITE_WITHOUT_RESP) {
+				inst->cli.desc_writable = true;
+			}
+		}
+
+		if (sub_params) {
+			sub_params->value = BT_GATT_CCC_NOTIFY;
+			sub_params->value_handle = chrc->value_handle;
+			/*
+			 * TODO: Don't assume that CCC is at handle + 2;
+			 * do proper discovery;
+			 */
+			sub_params->ccc_handle = attr->handle + 2;
+			sub_params->notify = aics_client_notify_handler;
+			bt_gatt_subscribe(conn, sub_params);
+		}
+	}
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static void aics_client_reset(struct bt_aics *inst, struct bt_conn *conn)
+{
+	inst->cli.desc_writable = 0;
+	inst->cli.change_counter = 0;
+	inst->cli.gain_mode = 0;
+	inst->cli.start_handle = 0;
+	inst->cli.end_handle = 0;
+	inst->cli.state_handle = 0;
+	inst->cli.gain_handle = 0;
+	inst->cli.type_handle = 0;
+	inst->cli.status_handle = 0;
+	inst->cli.control_handle = 0;
+	inst->cli.desc_handle = 0;
+
+	/* It's okay if these fail */
+	(void)bt_gatt_unsubscribe(conn, &inst->cli.state_sub_params);
+	(void)bt_gatt_unsubscribe(conn, &inst->cli.status_sub_params);
+	(void)bt_gatt_unsubscribe(conn, &inst->cli.desc_sub_params);
+}
+
+
+int bt_aics_discover(struct bt_conn *conn, struct bt_aics *inst,
+		     const struct bt_aics_discover_param *param)
+{
+	int err = 0;
+
+	CHECKIF(!inst || !conn || !param) {
+		BT_DBG("%s cannot be NULL",
+		       inst == NULL ? "inst" : conn == NULL ? "conn" : "param");
+		return -EINVAL;
+	}
+
+	CHECKIF(param->end_handle < param->start_handle) {
+		BT_DBG("start_handle (%u) shall be less than end_handle (%u)",
+		       param->start_handle, param->end_handle);
```

Comment:
Less than or equal?

Reply:
Less than is correct: We cannot do discovery on a single handle value (at least not in Zephyr as we have a check for that in `bt_gatt_discover`).

Reply:
The CHECKIF checks that the end handle is not smaller than the start handle.  So this allows the handles to be equal.
But the comment says that the start handle must be less than the end handle (equal is not allowed).
The correct thing to do seems to be to replace the "<" with a "<=".

Reply:
Oh, you are absolutely right; I too often think of `CHECKIF(cond)` instead of `CHECKF(err_cond)` which is actually the case. Thanks!